torch_xla/distributed/xla_dist.py

```diff
@@ -86,14 +102,39 @@ def __init__(self,
                ' will interfere with the values set for distributed'
                ' training'.format(dist_var)))
 
+  def _check_client_mesh_health(self, health_timeout=120):
+    max_delay = 0
+    now = time.time()
+    master_last = self._last_heartbeats[self._client_master.get_internal_ip()]
+    for cw_ip in self._last_heartbeats.keys():
+      if cw_ip == self._client_master.get_internal_ip():
+        continue
+      max_delay = max(
+        max_delay, master_last - self._last_heartbeats[cw_ip])
+
+    if max_delay > health_timeout:
+      self._error_queue.put(RuntimeError(
+        'Client mesh is unhealthy due to non-matching heartbeats'))
+
   def _stream_logs(self, process, client_worker):
-    client_ip = client_worker._internal_ip
-    ordinal = self._cluster._client_workers.index(client_worker)
+    client_ip = client_worker.get_internal_ip()
+    ordinal = self._cluster.get_client_workers().index(client_worker)
 
     def _stream_output(stream, log_fn):
       for std in iter(stream.readline, b''):
+        std_line = std.decode('utf-8').rstrip('\n')
+        if 'torch_xla.core.xla_model::mark_step' in std_line:
+          self._last_heartbeats[client_ip] = time.time()
+          # Only run expensive hang-check every N seconds.
+          time_since_last_check = time.time() - self._last_heartbeat_check_time
+          if client_ip == self._client_master.get_internal_ip() and time_since_last_check > self.HEARTBEAT_CHECK_PERIOD:
+            self._last_heartbeat_check_time = time.time()
+            t = threading.Thread(target=self._check_client_mesh_health)
```

Comment:
I think it should be safe to assume python >= 3.3 and use daemon=True as constructor argument.